File: src/main/java/com/telegram/folobot/service/handlers/ContextIndependentHandler.java
package com.telegram.folobot.service.handlers;

import com.telegram.folobot.ChatId;
import com.telegram.folobot.dto.FoloPidorDto;
import com.telegram.folobot.service.FoloPidorService;
import com.telegram.folobot.service.FoloUserService;
import com.telegram.folobot.service.MessageService;
import com.telegram.folobot.service.UserService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.meta.api.objects.Message;
import org.telegram.telegrambots.meta.api.objects.Update;
import org.telegram.telegrambots.meta.api.objects.User;

import java.util.Objects;

import static com.telegram.folobot.ChatId.isAndrew;

@Component
@RequiredArgsConstructor
public class ContextIndependentHandler {
    private final FoloUserService foloUserService;
    private final FoloPidorService foloPidorService;
    private final MessageService messageService;
    private final UserService userService;

    public void handle(Update update) {
        //Добавление фолопользователя в бд
        addFoloUser(update);

        //Пересылка личных сообщений в спецчат
        forwardPrivate(update);
    }

    /**
     * Залоггировать пользователя
     *
     * @param update {@link Update}
     */
    private void addFoloUser(Update update) {
        Message message = update.getMessage();
        if (Objects.isNull(message.getIsAutomaticForward()) || !message.getIsAutomaticForward()) {
            User user = message.getFrom();
            if (Objects.isNull(user) && !message.getNewChatMembers().isEmpty()) {
                user = message.getNewChatMembers().get(0);
            }
            if (!Objects.isNull(user)) {
                // Фолопользователь
                foloUserService.save(
                        foloUserService.findById(user.getId())
                                .setName(userService.getUserName(user)));
                // И фолопидор
                if (!message.isUserMessage() && !foloPidorService.existsById(message.getChatId(), user.getId())) {
                    foloPidorService.save(new FoloPidorDto(message.getChatId(), user.getId()));
                }
            }
        }
    }

    /**
     * Пересылка личных сообщений
     *
     * @param update {@link Update}
     */
    private void forwardPrivate(Update update) {
        if (update.hasMessage())
            if (update.getMessage().isUserMessage()) {
                messageService.forwardMessage(ChatId.getPOC_ID(), update);
            } else if (isAndrew(update.getMessage().getFrom())) {
                messageService.forwardMessage(ChatId.getANDREWSLEGACY_ID(), update);
            }
    }
}

File: src/main/kotlin/com/telegram/folobot/service/handlers/ContextIndependentHandler.kt
package com.telegram.folobot.service.handlers

import com.telegram.folobot.IdUtils
import com.telegram.folobot.IdUtils.Companion.isAndrew
import com.telegram.folobot.service.*
import mu.KLogging
import org.springframework.stereotype.Component
import org.telegram.telegrambots.meta.api.objects.Update

@Component
class ContextIndependentHandler(
    private val foloUserService: FoloUserService,
    private val foloPidorService: FoloPidorService,
    private val messageService: MessageService,
    private val foloIndexService: FoloIndexService
) : KLogging() {

    fun handle(update: Update) {
        //Добавление фолопользователя в бд
        saveFoloUser(update)

        //Пересылка личных сообщений в спецчат
        forwardPrivate(update)

        //Добавление очков активности
        addActivityPoints(update)
    }

    /**
     * Залоггировать пользователя
     *
     * @param update [Update]
     */
    private fun saveFoloUser(update: Update) {
        val message = update.message
        if (message.isAutomaticForward == null || !message.isAutomaticForward) {
            (message.from ?: message.newChatMembers?.firstOrNull())?.run {
                // Фолопользователь
                foloUserService.save(foloUserService.findById(this.id).setName(this.getName()))
                // И фолопидор
                if (!message.isUserMessage) {
                    foloPidorService.save(foloPidorService.findById(message.chatId, this.id).updateMessagesPerDay())
                }
                logger.info { "Saved foloUser ${this.getName()}" }
            }
        }
    }

    /**
     * Пересылка личных сообщений
     *
     * @param update [Update]
     */
    private fun forwardPrivate(update: Update) {
        if (update.hasMessage()) if (update.message.isUserMessage) {
            messageService.forwardMessage(IdUtils.POC_ID, update)
            logger.info { "Forwarded message to POC" }
        } else if (isAndrew(update.message.from)) {
            messageService.forwardMessage(IdUtils.ANDREWSLEGACY_ID, update)
            logger.info { "Forwarded message to Andrews legacy" }
        }
    }

    private fun addActivityPoints(update: Update) {
        foloIndexService.addActivityPoints(update)
    }
}